### ISSUE TITLE
intersection types and conditional types

### ISSUE THREAD

**Issue by marcosh** [vimeo/psalm#6942]
consider https://psalm.dev/r/8e871b2851

two things:
- the parser forces me to use parentheses around the intersection type. I would not expect that to be needed
- since from the condition we know that `A` is an `array`, I don't see why `A&array` should be forbidden

**Comment by orklah** [comment]
Why would `A&array` would be needed given that `A is array` already?

**Comment by marcosh** [comment]
@orklah I might be missing something here, I'm not 100% sure how `is` works in conditionals, but I would expect that `int[]` would satisfy `is array` and so `A & array` would be `int[]`.

I could probably use `A` directly, but the issue still remains

**Comment by orklah** [comment]
Yeah, what I meant, is basically, `(A is array ? (A&array) : null)` would be strictly equivalent to `(A is array ? A : null)`, unless there is a use case I don't see

**Comment by marcosh** [comment]
yes, they should be equivalent

**Comment by marcosh** [comment]
@orklah actually there are cases when this is actually an issue, see https://psalm.dev/r/779ffcc063

if the `A` template is, for example `array<int>`, I would to say that the return value is `non-empty-array<int>`

**Comment by orklah** [comment]
This should be strictly equivalent to this: https://psalm.dev/r/c036b61d71

**Comment by marcosh** [comment]
@orklah I'm not sure it is the same.

In your example you already know that the type of `$b` is `array{""}`.

in this other example https://psalm.dev/r/a69962b672, I would like Psalm to infer that the type of `$c` is `non-empty-array<array-key, int>|null` and not `array<array-key, int>|null`

**Comment by orklah** [comment]
Ok, I get it. You're trying to use a mix between conditional returns and assertions. (Your goal is to either make sure your input type is what you want (and return it) OR change the return type according to the behavior of the function.

It's quite unorthodox. If you have control over this code, I'd strongly advise to use assertions (something like https://psalm.dev/r/1ad50841b5)

or restrict a bit the param type to be able to restrict possibilities (something like https://psalm.dev/r/e821587e3f)

Otherwise, you're in a weird position where you have `A of mixed` in input, (so it could match for exemple the empty string) and then you declare `A&non-empty-array` to be the returned type.

When Psalm will try to check this, it will have to report that `A of mixed`, when applied the assertion `!empty` is still `A of mixed`, and not  `A&non-empty-array`. So you'd be left with an error anyway.

With that said, this syntax would not be out of place if it was supported. If you want to propose a PR for something like that and it seems coherent (meaning, as I said, that you should end up with an error in your specific example), then I'd consider merging that.

**Comment by marcosh** [comment]
Thanks for the detailed answer.

> Ok, I get it. You're trying to use a mix between conditional returns and assertions. (Your goal is to either make sure your input type is what you want (and return it) OR change the return type according to the behavior of the function.

My goal is to write a `notEmpty` validator, where a validator is something of type `A -> Either<E, B>` (`A` is the raw data, `E` is the message if the validation fails and `B` is the result of the validation).
Since the `empty` function in PHP accepts anything, I'd like the type `A` to be completely generic (otherwise I could have `notEmptyArray`, `notEmptyString`, `notEmptyAnythingElse`) and I would like the return type to be as precise as possible.

What I actually have at moment is more complicated that what I provided above in the issue and looks like

```php
/**
     * @template C
     * @template F
     * @param F $e
     * @return (C is string ? Validation<C, F, non-empty-string> : (C is array ? Validation<C, F, non-empty-array> : Validation<C, F, C>))
     */
    public static function notEmpty($e): self
```

The `string` case is easy, since there are no other string subtypes than `non-empty-string`. With `array` the situation is more complicated.

> Otherwise, you're in a weird position where you have `A of mixed` in input, (so it could match for exemple the empty string) and then you declare `A&non-empty-array` to be the returned type.
> 
> When Psalm will try to check this, it will have to report that `A of mixed`, when applied the assertion `!empty` is still `A of mixed`, and not `A&non-empty-array`. So you'd be left with an error anyway.

Understood. Still, I'm saying that the return type is `A&non-empty-array` only if `A is array`.
Would it help if I had an `is_array` check in there?

> With that said, this syntax would not be out of place if it was supported. If you want to propose a PR for something like that and it seems coherent (meaning, as I said, that you should end up with an error in your specific example), then I'd consider merging that.

Last time I checked the Psalm codebase I felt quite lost. Could you provide any pointer to where I should implement this?

**Comment by orklah** [comment]
> My goal is to write a notEmpty validator

Could you post a snippet of what it would look like? I'm not sure I see the difference with the assertion example I posted above but I must be missing something

> The string case is easy, since there are no other string subtypes than non-empty-string

Within Psalm type system, there are. (`non-empty-lowercase-string` or `non-empty-literal-string`) but you may not care about those

> Would it help if I had an is_array check in there?

I'm not really sure TBH

> Could you provide any pointer to where I should implement this?

If I'm right, your issue is here:
https://github.com/vimeo/psalm/blob/b29737f2b154748430250effdbcf47de8a5904df/src/Psalm/Internal/Type/TypeParser.php#L936
We're at a place where we try to process the intersections. Psalm allow intersection between objects OR intersection between arrays. I think your issue is that the Generic type under the hood of the Conditional Return is not considered as an array and Psalm refuse to do anything with it.

If you can help Psalm seeing that this is an array, I think you should be able to make some progress here

**Comment by weirdan** [comment]
> The string case is easy, since there are no other string subtypes than non-empty-string

Note that `'0'` is a `non-empty-string` to Psalm, but it's considered empty by PHP. There's a `non-falsy-string` type that matches PHP's `empty()` semantics.

**Comment by marcosh** [comment]
> Could you post a snippet of what it would look like? I'm not sure I see the difference with the assertion example I posted above but I must be missing something

You can find it [here](https://github.com/marcosh/lamphpda-validation/blob/d9621fcd3de9d81a89c213d85afedb24a06215f6/src/Validation.php#L139)
 
> If I'm right, your issue is here:
> 
> https://github.com/vimeo/psalm/blob/b29737f2b154748430250effdbcf47de8a5904df/src/Psalm/Internal/Type/TypeParser.php#L936
> 
> We're at a place where we try to process the intersections. Psalm allow intersection between objects OR intersection between arrays. I think your issue is that the Generic type under the hood of the Conditional Return is not considered as an array and Psalm refuse to do anything with it.
> 
> If you can help Psalm seeing that this is an array, I think you should be able to make some progress here

Thanks for the pointer! I'll try to take a look at it

**Comment by marcosh** [comment]
> > The string case is easy, since there are no other string subtypes than non-empty-string
> 
> Note that `'0'` is a `non-empty-string` to Psalm, but it's considered empty by PHP. There's a `non-falsy-string` type that matches PHP's `empty()` semantics.

oh, thanks for mentioning this!
